internal/cmd/launcher.go

```diff
@@ -0,0 +1,348 @@
+package cmd
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/grafana/k6deps"
+	"github.com/grafana/k6provider"
+	"go.k6.io/k6/cloudapi"
+	"go.k6.io/k6/cmd/state"
+	"go.k6.io/k6/internal/build"
+)
+
+var (
+	errScriptNotFound     = errors.New("script not found")
+	errUnsupportedFeature = errors.New("not supported")
+)
+
+// commandExecutor executes the requested k6 command line command.
+// It abstract the execution path from the concrete binary.
+type commandExecutor interface {
+	run(*state.GlobalState)
+}
+
+// Launcher is a k6 launcher. It analyses the requirements of a k6 execution,
+// then if required, it provisions a binary executor to satisfy the requirements.
+type Launcher struct {
+	// gs is the global state of k6.
+	gs *state.GlobalState
+
+	// provision generates a custom binary from the received list of dependencies
+	// with their constrains, and it returns an executor that satisfies them.
+	provision func(*state.GlobalState, k6deps.Dependencies) (commandExecutor, error)
```

Comment:
Nit (feel free to ignore/dismiss): although it is not strictly necessary here, I tend to prefer, when possible, defining and implementing an interface (`Provisioner`?), instead of a callback. The reason is ease of debugging and codebase navigation in IDEs. It's much easier to go to definition of an interface and ask for implementations, than to track down where the concrete callback instance is defined and passed down. Sometimes it also makes testing more straightforward.

No arguing on my end, really just a suggestion and expressing a preference, I'm okay with it as-is 👍

Reply:
@pablochacin I agree with @oleiade, at least it should be worth to have a reflection/attempt on refactor it. Can you open and add an issue to the stability epic, please?